lib/CodeGen/CGExpr.cpp

```diff
@@ -4306,3 +4310,41 @@ RValue CodeGenFunction::EmitPseudoObjectRValue(const PseudoObjectExpr *E,
 LValue CodeGenFunction::EmitPseudoObjectLValue(const PseudoObjectExpr *E) {
   return emitPseudoObjectExpr(*this, E, true, AggValueSlot::ignored()).LV;
 }
+
+llvm::Value *CodeGenFunction::EmitDynamicBoundsCast(CastExpr *CE) {
+  Expr *E = CE->getSubExpr();
+  QualType DestTy = CE->getType();
+  CastKind Kind = CE->getCastKind();
+  // Bounds casting has two functionalities
+  // - code generation for explicit type casting
+  // - code generation for dynamic check, dynamic_bounds_cast only
+  //  - dynamic_check(e1 != NULL)
+  //  - dynamic_check(lb <= e2 && e3 <= ub)
+  Address Addr = Address::invalid();
+  // For count bounds, source can be a pointer/array type (ArrayToPointerDecay)
+  // For range bounds, source can be a pointer/array/integer
+  // For integer type,
+  if (E->getType()->isPointerType()) {
+    Addr = EmitPointerWithAlignment(E);
+    // explicit type casting for destination type
+    Addr = Builder.CreateBitCast(Addr, ConvertType(DestTy));
+  } else {
+    // CK_IntegralToPointer (IntToPtr) casts integer to pointer type
+    llvm::Value *Src = EmitScalarExpr(const_cast<Expr *>(E));
+    llvm::Type *DestLLVMTy = ConvertType(DestTy);
+    llvm::Type *MiddleTy = CGM.getDataLayout().getIntPtrType(DestLLVMTy);
+    bool InputSigned = E->getType()->isSignedIntegerOrEnumerationType();
+    llvm::Value* IntResult =
+      Builder.CreateIntCast(Src, MiddleTy, InputSigned, "conv");
+    llvm::Value *Result = Builder.CreateIntToPtr(IntResult, DestLLVMTy);
+    CharUnits Align = getContext().getTypeAlignInChars(DestTy);
+    Addr = Address(Result, Align);
+  }
+  if (Kind == CK_DynamicPtrBounds) {
+    BoundsCastExpr *BCE = cast<BoundsCastExpr>(CE);
+    EmitDynamicNonNullCheck(Addr, E->getType());
```

Comment:
Let CastBounds be bounds (castlb, castub)
Let SubExprBounds be bounds (lb, ub)
I have designed to emit as follows:
```
dynamic_check(E != NULL);
dynamic_check(lb <= castlb && castub <= ub);
```
However, you mean that it is wrong and it SHOULD emit as follows:
`dynamic_check (E == NULL || (lb <= castlb && castub <= ub));`

Is it correct?
Can you explain why dynamic_check is skipped if E is NULL?